include/envoy/router/router.h

```diff
@@ -265,6 +274,13 @@ class VirtualHost {
    * nullptr is returned.
    */
   virtual const Protobuf::Message* perFilterConfig(const std::string& name) const PURE;
+
+  /**
+   * @return const PerRouteConfigObject* the per-filter config pre-processed object for
+   *  the given filter name. If not per-filter config for the filter, or the filter factory
+   *  doesn't create such object, nullptr is returned.
+   */
+  virtual const PerRouteConfigObject* perFilterConfigObject(const std::string& name) const PURE;
```

Comment:
I think we should add an additional template method:
```
template <typename Derived>
const Derived* perFilterConfigObjectTyped(const std::string& name) const {
    const PerRouteConfigObject* obj = perFilterConfigObject(name);
    if (obj == nullptr) {
        return obj;
    }
    ASSERT(dynamic_cast<const Derived*>(obj) != nullptr);
    return static_cast<const Derived*>(obj);
}
```

Or similar.

Reply:
Yeah we have basically this for TLS. +1